src/canopy/knowledge_base/record_encoder/hybrid.py

```diff
@@ -0,0 +1,136 @@
+import logging
+from functools import cached_property
+from typing import List, Optional
+
+from pinecone_text.hybrid import hybrid_convex_scale
+from pinecone_text.sparse import BM25Encoder
+
+from . import DenseRecordEncoder, OpenAIRecordEncoder
+from .base import RecordEncoder
+from canopy.knowledge_base.models import KBQuery, KBEncodedDocChunk, KBDocChunk
+from canopy.models.data_models import Query
+
+logger = logging.getLogger(__name__)
+
+
+class HybridRecordEncoder(RecordEncoder):
+    """
+    HybridRecordEncoder is a subclass of RecordEncoder that generates sparse and dense vector representation of
+    documents` chunks and textual queries.
+
+    The dense representation generated by the `HybridRecordEncoder` is a list of floats in a given dimension.
+    The sparse representation generated by the `HybridRecordEncoder` is a `SparseVector`.
+
+    HybridRecordEncoder uses BaseDenseEncoder for dense encoding and BM25Encoder for sparse encoding.
+
+    Alpha is a parameter that controls the weight of the dense vector in the hybrid representation.
+    If alpha is 1, the query vector will be the dense vector. The default value of alpha is 0.5.
+
+    For more information about the encoders see: https://github.com/pinecone-io/pinecone-text
+
+    """  # noqa: E501
+
+    _DEFAULT_COMPONENTS = {
+        "dense_record_encoder": OpenAIRecordEncoder
+    }
+
+    def __init__(self,
+                 dense_record_encoder: Optional[DenseRecordEncoder] = None,
+                 alpha: float = 0.5,
+                 bm_25_encoder_df_path: Optional[str] = None,
+                 **kwargs):
+        """
+        Initialize the encoder.
+
+        Args:
+            dense_record_encoder: A DenseRecordEncoder to encode the text.
+            alpha: The weight of the dense vector in the hybrid representation (between 0 and 1).
+            bm_25_encoder_df_path: The path to the file that contains the document frequencies of the BM25Encoder.\
+            You can create this file by fitting the BM25Encoder on a corpus of documents and calling `dump`\
+            on the encoder.
+            **kwargs: Additional arguments to pass to the RecordEncoder.
+        """  # noqa: E501
+
+        if not 0 < alpha <= 1:
```

Comment:
We should probably add an explicit error message for `alpha==0`, saying "sparse only not supported"

Reply:
Added